### DISCUSSION TITLE
Support asyncpg copy_records_to_table

### DISCUSSION THREAD

**Discussion by vladjohnson** [sqlalchemy/sqlalchemy#11711] [Usage Questions] [answered]
### Describe the use case

copy_records_to_table method of asyncpg driver is incredibly performant -- allowing to insert almost a million rows per second on a single connection. SQLAlchemy users would significantly benefit from implementation of this function, similarly to psycopg3's copy_expert.

### Databases / Backends / Drivers targeted

Database: Postgres
Drivers: asyncpg

### Example Use

Similarly to copy_expert, we'd expect:

```python
raw_connection = engine.raw_connection()
raw_cursor = raw_connection.cursor()
raw_cursor.copy_record_to_table(...)
```

**Comment by CaselIT** [comment]
Hi,

You have access to the asyncpg connection using driver_connection. See https://docs.sqlalchemy.org/en/20/core/pooling.html#sqlalchemy.pool.PoolProxiedConnection.driver_connection

**Reply by vladjohnson** [reply]
Hey @CaselIT thanks for responding so quickly. Can you please give me an example of how to use copy_record_to_table with the driver_connection? As far as I saw in the dialects, it is not supported. Thanks

**Reply by zzzeek** [reply, marked as answer]
from the link above, it follows into the FAQ where we have https://docs.sqlalchemy.org/en/20/faq/connections.html#accessing-the-underlying-connection-for-an-asyncio-driver  so I mix that code example with the one from https://magicstack.github.io/asyncpg/current/api/index.html#asyncpg.connection.Connection.copy_records_to_table


```py
async def main():
    engine = create_async_engine(...)
    conn = await engine.connect()

    # pep-249 style ConnectionFairy connection pool proxy object
    # presents a sync interface
    connection_fairy = await conn.get_raw_connection()

    # the really-real innermost driver connection is available
    # from the .driver_connection attribute
    raw_asyncio_connection = connection_fairy.driver_connection

    # work with raw asyncio connection

    result = await raw_asyncio_connection.copy_records_to_table(

        'mytable', records=[

            (1, 'foo', 'bar'),

            (2, 'ham', 'spam')])
```

**Reply by vladjohnson** [reply]
Thank you so much, Michael! Appreciate this answer and all of your work.
